Update helm.md (#33307)
sourcegraph/sourcegraph: doc/admin/install/kubernetes/helm.md

# Using Helm

## Requirements

* [Helm 3 CLI](https://helm.sh/docs/intro/install/)
* Kubernetes 1.19 or greater

<div class="cta-group">
<!--<a class="btn btn-primary" href="#installation">★ ...</a>-->

> <a class="btn" href="#configuration">Configuration</a>

Author: ```suggestion
<a class="btn" href="#configuration">Configuration</a>
- <a class="btn" href="#configure-sourcegraph-on-google-kubernetes-engine-gke">Google GKE</a>
```

<a class="btn" href="#configure-sourcegraph-on-elastic-kubernetes-service-eks">Google GKE</a>
<a class="btn" href="#configure-sourcegraph-on-azure-managed-kubernetes-service-aks">Azure AKS</a>
<a class="btn" href="#configure-sourcegraph-on-other-cloud-providers-or-on-prem">Other cloud providers or on-prem</a>
<a class="btn" href="#upgrading-sourcegraph">Upgrading</a>
</div>

> ## High level overview of how to use Helm with Sourcegraph

Author: Is the advice in this new section legit?

Is prepping customizations before starting a good idea?

1. Prepare any required customizations
   - Sourcegraph can be deployed using Helm with no customizations to the default [values.yaml](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml) but there are a range of [supported options](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph#configuration-options). 
   - It's worth reviewing these and preparing them ahead of anything else, using the guidance in [Configuration](#configuration).
1. Review the changes
   - There are [three mechanisms](#reviewing-changes) that can be used to review any customizations made, this is an optional step, but may be useful the first time you deploy Sourcegraph, for peace of mind.

> 1. Select your deployment method and follow the guidance:

Author: ```suggestion
1. Select your deployment method and follow the guidance:
   - [Google GKE](#configure-sourcegraph-on-google-kubernetes-engine-gke)
```

>    - [Google GKE](#configure-sourcegraph-on-elastic-kubernetes-service-eks)

Reviewer: ```suggestion
   - [AWS EKS](#configure-sourcegraph-on-elastic-kubernetes-service-eks)
```
Same as above - we're missing Google's section

   - [Azure AKS](#configure-sourcegraph-on-azure-managed-kubernetes-service-aks)
   - [Other cloud providers or on-prem](#configure-sourcegraph-on-other-cloud-providers-or-on-prem)

## Quickstart

> > ℹ️ This quickstart guide is useful to those already familiar with Helm who have a good unstanding of how to use Helm in the environment they want to deploy into, and just who want to quickly deploy Sourcegraph with Helm with defalt configuration. If this doesn't cover what you need to know, see the links above for platform specifit guides.

Reviewer: ```suggestion
> ℹ️ This quickstart guide is useful to those already familiar with Helm who have a good understanding of how to use Helm in the environment they want to deploy into, and just who want to quickly deploy Sourcegraph with Helm with default configuration. If this doesn't cover what you need to know, see the links above for platform specific guides.
```

(Although I liked your suggestion about getting rid of this section and just directing everyone to provider-specific instructions)

# Using Helm

## Requirements

* [Helm 3 CLI](https://helm.sh/docs/intro/install/)
* Kubernetes 1.19 or greater

<div class="cta-group">
<!--<a class="btn btn-primary" href="#installation">★ ...</a>-->
<a class="btn" href="#configuration">Configuration</a>
- <a class="btn" href="#configure-sourcegraph-on-google-kubernetes-engine-gke">Google GKE</a>
<a class="btn" href="#configure-sourcegraph-on-elastic-kubernetes-service-eks">AWS EKS</a>
<a class="btn" href="#configure-sourcegraph-on-azure-managed-kubernetes-service-aks">Azure AKS</a>
<a class="btn" href="#configure-sourcegraph-on-other-cloud-providers-or-on-prem">Other cloud providers or on-prem</a>
<a class="btn" href="#upgrading-sourcegraph">Upgrading</a>
</div>

## High level overview of how to use Helm with Sourcegraph

1. Prepare any required customizations

>    - Sourcegraph can be deployed using Helm with no customizations to the default [values.yaml](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml) but there are a range of [supported options](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph#configuration-options). 

Author: ```suggestion
```

>    - It's worth reviewing these and preparing them ahead of anything else, using the guidance in [Configuration](#configuration).

Author: ```suggestion
   - Most environment are likely to need changes from the defaults - use the guidance in [Configuration](#configuration).
```

helm install --version 0.7.0 sourcegraph sourcegraph/sourcegraph
```

Sourcegraph should now be available via the address set. Browsing to the url should now provide access to the Sourcegraph UI to create the initial administrator account. 

Now the deployment is complete, more information on configuring the Sourcegraph application can be found here:
[Configuring Sourcegraph](../../config)

## Configuration

The Sourcegraph Helm chart is highly customizable to support a wide range of environments. We highly recommend that customizations be applied using an override file, which allows customizations to persist through upgrades without needing to manage merge conflicts. 

The default configuration values can be viewed in the [values.yaml](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml) file along with all [supported options](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph#configuration-options). 

To customize configuration settings with an override file, create an empty yaml file (e.g. `override.yaml`) to get started. 

> (There are no pre-requisites to creating this file ahead of deployment!)

Reviewer: This is confusing - I think you're trying to convey that this file can be prepared in advance of the deployment?

> (There are no pre-requisites to creating this file ahead of deployment!)

Author: ```suggestion
(The configuration override file can be created in advance of deployment!)
```

#### Prerequisites

> 1. You need to have a GKE cluster (>=1.19) with the `HTTP Load Balancing` addon enabled:
> > Alternatively, you can use your own choice of Ingress Controller and disable the `HTTP Load Balancing` add-on, [learn more](https://cloud.google.com/kubernetes-engine/docs/how-to/custom-ingress-controller).

Reviewer 2: ```suggestion
1. You need to have a GKE cluster (>=1.19) with the `HTTP Load Balancing` addon enabled. Alternatively, you can use your own choice of Ingress Controller and disable the `HTTP Load Balancing` add-on, [learn more](https://cloud.google.com/kubernetes-engine/docs/how-to/custom-ingress-controller).
```

the formating was messed up

> 2. Your account should have sufficient access equivalent to the `cluster-admin` ClusterRole.
> 3. Connect to your cluster (via either the console or the command line using `gcloud`) and ensure the cluster is up and running by running: `kubectl get nodes` (several `ready` nodes should be listed)
> 4. Have the Helm CLI installed and run the following command to link to the Sourcegraph helm repository

Reviewer 2: ```suggestion
1. Your account should have sufficient access equivalent to the `cluster-admin` ClusterRole.
1. Connect to your cluster (via either the console or the command line using `gcloud`) and ensure the cluster is up and running by running: `kubectl get nodes` (several `ready` nodes should be listed)
1. Have the Helm CLI installed and run the following command to link to the Sourcegraph helm repository
```

another nit pick, can we change all sequence to `1`? Markdown will transform them to proper sequence number.

it's just easier because we don't have to count when adding more items to the list

```sh
helm repo add sourcegraph https://sourcegraph.github.io/deploy-sourcegraph-helm/
```

#### Steps

**1** – Create your override file and add in any configuration override settings you need - see [configuration](#configuration) for more information on override files and the options around what can be conifgured.

> Add into your overide file the below values to configure both your ingress hostname and your storage class. We recommend configuring Ingress to use [Container-native load balancing] to expose Sourcegraph publicly on a domain of your choosing and setting the Storage Class to use [Compute Engine persistent disk]. (For an example file see [override.yaml](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph/examples/gcp/override.yaml))

Reviewer 2: the intention of putting the link `[override.yaml]` directly above the snippet is provide us the ability to do something similar to

https://github.com/dagger/dagger/blob/main/docs/core-concepts/1204-secrets.md?plain=1
https://docs.dagger.io/1204/secrets#sops

in the future

where we can reference a remote file from the snippet from markdown, and it will render nicely

with the current approach, we will always risk values getting out-of-sync

maybe put it back?

> ```yaml

Author: ```suggestion
[override.yaml](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph/examples/gcp/override.yaml)
```yaml
```

      # kubernetes.io/ingress.global-static-ip-name: ADDRESS_NAME
```

If using another certificate, add the following to Ingress:
```

> TODO

Author: TODO

Reviewer 2: this step would be almost identical with https://github.com/sourcegraph/sourcegraph/blob/5f418ebc09629e51b608a9a37a0311454d1ed0f4/doc/admin/install/kubernetes/helm.md?plain=1#L558-L592

maybe we can extract this out to a standalone guide in the `#confugration` section, like `Using custom TLS secrets in Ingress`

Author: For now - can you just replicate it with any necessary changes please @michaellzc

      # kubernetes.io/ingress.global-static-ip-name: ADDRESS_NAME
```

If using another certificate, add the following to Ingress:
```
TODO
```
**5** – Validate the deployment
Sourcegraph should now be available via the address set. 
Browsing to the url should now provide access to the Sourcegraph UI to create the initial administrator account. 

**6** – Further configuration

Now the deployment is complete, more information on configuring the Sourcegraph application can be found here:
[Configuring Sourcegraph](../../config)

### Configure Sourcegraph on Elastic Kubernetes Service (EKS)

#### Prerequisites

1. You need to have a EKS cluster (>=1.19) with the following addons enabled:
   - [AWS Load Balancer Controller](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html)
   - [AWS EBS CSI driver](https://docs.aws.amazon.com/eks/latest/userguide/managing-ebs-csi.html)
> You may consider deploying your own Ingress Controller instead of ALB Ingress Controller, [learn more](https://kubernetes.github.io/ingress-nginx/)

> 2. Your account should have sufficient access equivalent to the `cluster-admin` ClusterRole.
> 3. Connect to your cluster (via either the console or the command line using `eksctl`) and ensure the cluster is up and running by running: `kubectl get nodes` (several `ready` nodes should be listed)
> 4. Have the Helm CLI installed and run the following command to link to the Sourcegraph helm repository

Reviewer 2: ```suggestion
1. Your account should have sufficient access equivalent to the `cluster-admin` ClusterRole.
1. Connect to your cluster (via either the console or the command line using `eksctl`) and ensure the cluster is up and running by running: `kubectl get nodes` (several `ready` nodes should be listed)
1. Have the Helm CLI installed and run the following command to link to the Sourcegraph helm repository
```

#### Prerequisites

> 1. You need to have a AKS cluster (>=1.19) with the following addons enabled:
>    - [Azure Application Gateway Ingress Controller](https://docs.microsoft.com/en-us/azure/application-gateway/ingress-controller-install-new)
>    - [Azure Disk CSI driver](https://docs.microsoft.com/en-us/azure/aks/csi-storage-drivers)
> > You may consider using your custom Ingress Controller instead of Application Gateway, [learn more](https://docs.microsoft.com/en-us/azure/aks/ingress-basic)

Reviewer 2: the formatting is messed up now. Should we put this above or below?

```suggestion
> You may consider using your custom Ingress Controller instead of Application Gateway, [learn more](https://docs.microsoft.com/en-us/azure/aks/ingress-basic)

1. You need to have a AKS cluster (>=1.19) with the following addons enabled:
   - [Azure Application Gateway Ingress Controller](https://docs.microsoft.com/en-us/azure/application-gateway/ingress-controller-install-new)
   - [Azure Disk CSI driver](https://docs.microsoft.com/en-us/azure/aks/csi-storage-drivers)
```

Author: I think below. Putting it above it lacks context. And it's still within that step.

2. Your account should have sufficient access equivalent to the `cluster-admin` ClusterRole.
3. Connect to your cluster (via either the console or the command line using the Azure CLI) and ensure the cluster is up and running by running: `kubectl get nodes` (several `ready` nodes should be listed)
4. Have the Helm CLI installed and run the following command to link to the Sourcegraph helm repository

```sh
helm repo add sourcegraph https://sourcegraph.github.io/deploy-sourcegraph-helm/
```

#### Steps

**1** – Create your override file and add in any configuration override settings you need - see [configuration](#configuration) for more information on override files and the options around what can be conifgured.

> Add into your overide file the below values to configure both your ingress hostname and your storage class. We recommend configuring Ingress to use [Application Gateway](https://azure.microsoft.com/en-us/services/application-gateway) to expose Sourcegraph publicly on a domain of your choosing and Storage Class to use [Azure Disk CSI driver](https://docs.microsoft.com/en-us/azure/aks/azure-disk-csi).

Reviewer: ```suggestion
Add into your override file the below values to configure both your ingress hostname and your storage class. We recommend configuring Ingress to use [Application Gateway](https://azure.microsoft.com/en-us/services/application-gateway) to expose Sourcegraph publicly on a domain of your choosing and Storage Class to use [Azure Disk CSI driver](https://docs.microsoft.com/en-us/azure/aks/azure-disk-csi).
```
